revisions/announcer/announcer.go

```diff
@@ -0,0 +1,293 @@
+	"gopkg.in/src-d/go-git.v4/config"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/object"
+	"gopkg.in/src-d/go-git.v4/plumbing/storer"
+	"gopkg.in/src-d/go-git.v4/storage/memory"
+)
+
+const mergedPrTagPrefix = "refs/tags/merge_pr_"
+
+var errNotAllEpochsConsumed = errors.New("Not all epochs consumed")
+var errNilRepo = errors.New("Repository may not be nil")
+var errVacuousEpochs = errors.New("[]epoch.Epoch slice is vacuous: contains no epochs")
+
+// Limits defines the start-time (lower bound) and now-time (upper bound) on a commit time-bounded query for revisions.
+type Limits struct {
+	Start time.Time
+	Now   time.Time
+}
+
+// ByCommitTimeDesc implements ordering object.Commit pointers by commit time, descending.
+type ByCommitTimeDesc []*object.Commit
+
+func (cs ByCommitTimeDesc) Len() int {
+	return len(cs)
+}
+func (cs ByCommitTimeDesc) Swap(i, j int) {
+	cs[i], cs[j] = cs[j], cs[i]
+}
+func (cs ByCommitTimeDesc) Less(i, j int) bool {
+	if cs[i] == nil {
+		return false
+	}
+	if cs[j] == nil {
+		return true
+	}
+	return cs[i].Committer.When.After(cs[j].Committer.When)
+}
+
+// GetErrNotAllEpochsConsumed produces the canonical error for failing to consume all input epochs.
+func GetErrNotAllEpochsConsumed() error {
+	return errNotAllEpochsConsumed
+}
+
+// GetErrNilRepo produces the canonical error for a nil repo value that was expected to be non-nil.
+func GetErrNilRepo() error {
+	return errNilRepo
+}
+
+// GetErrVacuousEpochs the canonical error for a vacuous computation over epochs; i.e., passing an empty slice of epochs which would yield an empty output.
+func GetErrVacuousEpochs() error {
+	return errVacuousEpochs
+}
+
+// EpochReferenceIterFactory is an interface for instantiating appropriate storer.ReferenceIter implementations for an announcer configuration.
+type EpochReferenceIterFactory interface {
+	GetIter(repo agit.Repository, limits Limits) (storer.ReferenceIter, error)
+}
+
+type boundedMergedPRIterFactory struct{}
+
+func (f boundedMergedPRIterFactory) GetIter(repo agit.Repository, limits Limits) (storer.ReferenceIter, error) {
+	if repo == nil {
+		return nil, errNilRepo
+	}
+
+	// (1) Start with all tags                                         [tagsIter]
+	// (2) Filter tags to included nothing but "merge_pr_*" tags, and  [prIter]
+	//     order by descending commit time                             [prIter]
+	// (3) Skip tags after basis.Now, and                              [return]
+	// (4) Stop iteration when commit time is before basis.Start       [return]
+
+	tagsIter, err := repo.Tags()
+	if err != nil {
+		log.Printf("ERRO: Failed to create git remote reference iter: %v", err)
+		return nil, err
+	}
+
+	prIter, err := agit.NewMergedPRIter(tagsIter, repo)
+	if err != nil {
+		log.Printf("ERRO: Failed to create git remote reference iter: %v", err)
+		return nil, err
+	}
+
+	return agit.NewStopReferenceIter(agit.NewStartReferenceIter(prIter, func(ref *plumbing.Reference) bool {
```

Comment:
nit: declare + name the 2 func parameters here, above, so this call is more readable.

Reply:
Done.